src/geo/transform.js

```diff
@@ -315,7 +315,7 @@ class Transform {
     }
 
     get cameraWorldSize(): number {
-        const distance = Math.max(this._camera.getDistanceToElevation(this._averageElevation), Number.EPSILON);
+        const distance = Math.max(this._camera.getDistanceToElevation(this._centerAltitude, this._center.lat), Number.EPSILON);
```

Comment:
I don't see any bugs in the implementation other than mercator coordinate being used instead of a latitude one. This change introduced here means that the camera world size is computed relative to camera center (which is exactly what the default worldSize does) which might vary a lot especially in mountainous areas. I haven't tried this out but I would think that the fog visual changes as a function of centre altitude which might not be desirable.

Reply:
cc @karimnaaji

Reply:
@mpulkki-mapbox it's not enough to only fix convert to mercator to lat/lng. `getDistanceToElevation` effectively returns the altitude of the camera from the center (not camera center) before taking pitch into account. My guess was that originally perhaps this was meant to be calculated as such. On the other hand as I mentioned in the PR description, simply using `worldSize` instead of `cameraWorldSize` to address this bug would not alter the behaviour of the fog at all.

Reply:
`getDistanceToElevation` is pretty much just a ray-plane intersection test where the physical camera position is the ray origin and `z = averageElevation` is the target plane. Changing `averageElevation` to `centerAltitude` is a fundamental change to the function and I'm interested in hearing why that's required to fix the original issue. Could you please update the ticket description to better describe what's actually fixed? :)

Reply:
@mpulkki-mapbox  exactly, it wasn't clear to me why we use the `averageElevation` instead of `centerAltitude`. I didn't try this out as I was fairly certain it shouldn't cause any break. But now I see from the video you provided that the fog indeed is jittering. The jittering is likely to occur anyways but is less noticeable because we do sampling for the camera center. For the map center however there's no such sampling leading to the jittering. Interestingly though for the symbols I still believe what's implemented here is the correct one since we don't really care about the elevation there anyways. Thanks again for taking time and carefully catching this one!
With this in mind I am going change the code back to use `averageElevation`. The fix here is then:
1. Fixes the missing conversion (mercator y to lat)
2. Uses `worldSize` for the `calculateDistanceTileData` to address the bug at hand.

Reply:
Hi 👋 This PR covers the reason for `averageElevation`: https://github.com/mapbox/mapbox-gl-js/pull/10621. The development of `averageElevation` was probably the most challenging and frustrating aspect of getting fog to work acceptably. The fundamental problem was that it's not exactly clear how to scale fog, as it's not tied to a static physical scale. At the same time, the map zoom is also unacceptable for scaling fog. Where 3D is concerned, the zoom—which _is_ used to scale line widths and other style properties—turns out not to work well at all, as it updates discontinuously and changes significantly, depending on what the centerline of the camera intersects.

So the question is what to use in order to determine a suitable scale for fog. If you intersect the camera ray with sea level, then at nonzero pitch and nonzero altitude, the distance to sea level is significantly farther than the distance to the terrain and fog disappears entirely. The effect then is that no matter how you configure fog, it always disappears as conditions change. We wanted to be able to configure it once and have it feel the same no matter where you are in the world.

If you use centerAltitude, then there are two problems: The first is that this only updates on map idle, so as you pan and whenever terrain is present, the scale isn't right and fog usually either obscures everything or disappears entirely. The second is that it's quite jumpy, as it only takes a single point into account (at the center of the frame) which doesn't accurately reflect the overall picture you're looking at.

Our solution was to sample five points, stretched to fit between the bottom of the map and the horizon line. The points are sampled at a rate of about twice per second whenever the map is being redraw. We compute the elevation of those five points and from that, the average elevation. Finally, the average elevation is time-eased so that updates are smooth. During rendering, we intersect the camera ray with a plane at that average elevation in order to scale the fog.

![116757286-a44a1480-a9c2-11eb-9675-7496317578d1](https://user-images.githubusercontent.com/572717/204368019-79abbfbe-9c38-467b-a048-6eb1ce6499dc.jpg)

I know that's a bit tricky, but I hope that at least explains the rationale for this complexity. Please let me know if any of the motivation is unclear.

Reply:
Here's a quick sketch of a side-view representation of what's going on. Black is the camera frustum. The blue X is what happens if you intersect the camera centerline with sea level: it's too far away and fog disappears. The red X is what happens if you intersect _only_ the centerline with terrain: it's very sensitive to bumpiness of terrain. The green X represents what happens if you average five points and intersect the camera centerline with a plane at that altitude: it's a better representation of what you're looking at which isn't so sensitive to bumps.

<img width="866" alt="Screen Shot 2022-11-28 at 12 17 00 PM" src="https://user-images.githubusercontent.com/572717/204372497-8400c6fe-3dbc-453c-bb6c-79ee789e7fdb.png">

Reply:
@rreusser Thanks a lot for this detailed explanation!

Reply:
Ah, yes, I think the camera moves freely in three dimensions during interaction, not directly in terms of the map center and zoom. But whenever the map idles, it ray-intersects the terrain, recomputes the camera center, and sets the zoom accordingly.